src/PreallocatedLinearOperators.jl

```diff
@@ -0,0 +1,135 @@
+  s *= @sprintf("  eltype: %s\n", eltype(op))
+  s *= @sprintf("  symmetric: %s\n", op.symmetric)
+  s *= @sprintf("  hermitian: %s\n", op.hermitian)
+  s *= "\n"
+  print(io, s)
+end
+
+# Constructors.
+"""
+    PreallocatedLinearOperator(Mv, Mtu, Maw, M; symmetric=false, hermitian=false)
+
+Construct a linear operator from a dense or sparse matrix, using `Mv` as storage space
+for `M * v` and `Mtu` as storage space for `transpose(M) * u` and `Maw` to store
+`adjoint(M) * w`. Use the optional keyword arguments to indicate whether the operator
+is symmetric and/or hermitian.
+"""
+function PreallocatedLinearOperator(Mv :: Vector{T}, Mtu :: Vector{T}, Maw :: Vector{T},
+                                    M :: AbstractMatrix{T};
+                                    symmetric=false, hermitian=false) where T
+  nrow, ncol = size(M)
+  @assert length(Mv) ≥ nrow
+  @assert length(Mtu) ≥ ncol
+  @assert length(Maw) ≥ ncol
+  prod = @closure v -> mul!(Mv, M, v)
+  tprod = @closure u -> mul!(Mtu, transpose(M), u)
+  ctprod = @closure w -> mul!(Maw, adjoint(M), w)
+  F1 = typeof(prod)
+  F2 = typeof(tprod)
+  F3 = typeof(ctprod)
+  PreallocatedLinearOperator{T,F1,F2,F3}(nrow, ncol, symmetric, hermitian, prod, tprod, ctprod)
+end
+
+"""
+    PreallocatedLinearOperator(Mv, M; symmetric=false, hermitian=false)
+
+Construct a linear operator with preallocation using `Mv` as storage space for all
+matrix-vector products.
+"""
+PreallocatedLinearOperator(Mv :: Vector{T}, M :: AbstractMatrix{T};
+                           symmetric=false, hermitian=false) where T =
+  PreallocatedLinearOperator(Mv, Mv, Mv, M, symmetric=symmetric, hermitian=hermitian)
+
+function PreallocatedLinearOperator(M :: AbstractMatrix{T};
+                                    symmetric=false, hermitian=false) where T
+  nrow, ncol = size(M)
+  local Mv, Mtu, Maw
+  if T <: Real
+    if symmetric
+      Maw = Mtu = Mv = Vector{T}(undef, nrow)
+    else
+      Mv = Vector{T}(undef, nrow)
+      Maw = Mtu = Vector{T}(undef, ncol)
+    end
+  else
+    if symmetric && hermitian # Actually real, but T is not
+      Maw = Mtu = Mv = Vector{T}(undef, nrow)
+    elseif symmetric
+      Mtu = Mv = Vector{T}(undef, nrow)
+      Maw = Vector{T}(undef, ncol)
+    elseif symmetric
+      Mv = Vector{T}(undef, nrow)
+      Maw = Mtu = Vector{T}(undef, ncol)
+    else
+      Mv = Vector{T}(undef, nrow)
+      Mtu = Vector{T}(undef, ncol)
+      Maw = Vector{T}(undef, ncol)
+    end
+  end
+  PreallocatedLinearOperator(Mv, Mtu, Maw, symmetric=symmetric, hermitian=hermitian)
+end
+
+"""
+    PreallocatedLinearOperator(M)
+
+Constructs a linear operator from a symmetric tridiagonal matrix. If
+its elements are real, it is also Hermitian, otherwise complex
+symmetric.
+"""
+function PreallocatedLinearOperator(M :: SymTridiagonal{T}) where T
+  nrow = size(M, 1)
+  Mv = Vector{T}(undef, nrow)
+  PreallocatedLinearOperator(Mv, Mv, Mv, M, symmetric=true, hermitian=eltype(M) <: Real)
```

Comment:
Should it really be 3 times the same storage? If M is symmetric but not hermitian, shouldn't we preallocate two vectors?

Reply:
You're right, thanks.